File: Iveely.Framework/Iveely.Framework/Text/Html2Article.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Iveely.Framework.Text
{
    public class Html2Article
    {
        public class ArticleDocument
        {
            public string Url { get; set; }

            public string Site { get; set; }

            public string Title { get; set; }

            /// <summary>
            /// 正文文本
            /// </summary>
            public string Content { get; set; }

            /// <summary>
            /// 带标签正文
            /// </summary>
            public string ContentWithTags { get; set; }

            /// <summary>
            /// 发布日期
            /// </summary>
            public DateTime PublishDate { get; set; }

            public string Timestamp { get; set; }

            /// <summary>
            /// 包含的字链接
            /// </summary>
            public List<string> ChildrenLink { get; set; }

            public override string ToString()
            {
                return Url + "\t" + Site + "\t" + Title + "\t" + PublishDate.ToShortDateString() + "\t" + Content + "\t" +
                       ContentWithTags;
            }
        }


        #region 参数设置

        // 正则表达式过滤：正则表达式，要替换成的文本
        private static readonly string[][] Filters = new string[][]
        {
            new string[] {@"(?is)<script.*?>.*?</script>", ""},
            new string[] {@"(?is)<style.*?>.*?</style>", ""},
            // 针对链接密集型的网站的处理，主要是门户类的网站，降低链接干扰
            new string[] {@"(?is)</a>", "</a>\n"}
        };

        private static bool _appendMode = false;

        /// <summary>
        /// 是否使用追加模式，默认为false
        /// 使用追加模式后，会将符合过滤条件的所有文本提取出来
        /// </summary>
        public static bool AppendMode
        {
            get { return _appendMode; }
            set { _appendMode = value; }
        }

        private static int _depth = 6;

        /// <summary>
        /// 按行分析的深度，默认为6
        /// </summary>
        public static int Depth
        {
            get { return _depth; }
            set { _depth = value; }
        }

        private static int _limitCount = 180;

        /// <summary>
        /// 字符限定数，当分析的文本数量达到限定数则认为进入正文内容
        /// 默认180个字符数
        /// </summary>
        public static int LimitCount
        {
            get { return _limitCount; }
            set { _limitCount = value; }
        }

        // 确定文章正文头部时，向上查找，连续的空行到达_headEmptyLines，则停止查找
        private static int _headEmptyLines = 2;
        // 用于确定文章结束的字符数
        private static int _endLimitCharCount = 20;

        #endregion


        public static ArticleDocument GetArticle(string url, ref bool isGetContentSuccess)
        {
            try
            {
                Html html = Html.CreatHtml(new Uri(url));
                if (html == null)
                {
                    return null;
                }
                ArticleDocument doc = GetRequestArticle(html.SourceCode);
                isGetContentSuccess = doc.Content.Length > 0;
                if (doc != null)
                {
                    doc.Url = url;
                    doc.Content = doc.Content.Length > 0 ? doc.Content : html.Content;
                    doc.Timestamp = html.Timestamp;
                    doc.Site = GetSiteName(html.Title);
                    doc.ChildrenLink = html.ChildrenLink.Select(o => o.ToString()).ToList();

                }

                return doc;
            }
            catch (Exception exception)
            {
                Console.WriteLine(exception.Message);
            }
            return null;
        }

        private static string GetSiteName(string siteTitle)
        {
            if (siteTitle.Contains("-"))
            {
                string[] infos = siteTitle.Split(new char[] { '-', '_', ' ' }, StringSplitOptions.RemoveEmptyEntries);
                if (infos.Length >= 1)
                {
                    return infos[infos.Length - 1];
                }
                return siteTitle;
            }
            else
            {
                return siteTitle;
            }
        }

        /// <summary>
        /// 从给定的Html原始文本中获取正文信息
        /// </summary>
        /// <param name="html"></param>
        /// <returns></returns>
        private static ArticleDocument GetRequestArticle(string html)
        {
            // 如果换行符的数量小于10，则认为html为压缩后的html
            // 由于处理算法是按照行进行处理，需要为html标签添加换行符，便于处理
            if (html.Count(c => c == '\n') < 10)
            {
                html = html.Replace(">", ">\n");
            }

            // 获取html，body标签内容
            string body = "";
            string bodyFilter = @"(?is)<body.*?</body>";
            Match m = Regex.Match(html, bodyFilter);
            if (m.Success)
            {
                body = m.ToString();
            }
            // 过滤样式，脚本等不相干标签
            foreach (var filter in Html2Article.Filters)
            {
                body = Regex.Replace(body, filter[0], filter[1]);
            }
            // 标签规整化处理，将标签属性格式化处理到同一行
            // 处理形如以下的标签：
            //  <a 
            //   href='http://www.baidu.com'
            //   class='test'
            // 处理后为
            //  <a href='http://www.baidu.com' class='test'>
            body = Regex.Replace(body, @"(<[^<>]+)\s*\n\s*", FormatTag);

            string content;
            string contentWithTags;
            string title1;
            GetContent(body, out content, out title1, out contentWithTags);
            string title2 = GetTitle(html);
            string title = GetMostProTitle(title1, title2, content);
            ArticleDocument article = new ArticleDocument
            {
                Title = title,
                PublishDate = GetPublishDate(html),
                Content = content,
                ContentWithTags = contentWithTags
            };

            return article;
        }

        /// <summary>
        /// 获取最大可能的标题
        /// </summary>
        /// <param name="title1"></param>
        /// <param name="title2"></param>
        /// <param name="body"></param>
        /// <returns></returns>
        private static string GetMostProTitle(string title1, string title2, string body)
        {
            // 如果任意一个为empty，则返回另外一个
            if (string.IsNullOrEmpty(title1)) { return title2; }
            if (string.IsNullOrEmpty(title2)) { return title1; }

            // 计算title1积分
            char[] titleArray = title1.ToArray();
            int count1 = 0;
            foreach (char c in titleArray)
            {
                if (body.Contains(c.ToString()))
                {
                    count1++;
                }
            }

            // 计算title2积分
            titleArray = title2.ToArray();
            int count2 = 0;
            foreach (char c in titleArray)
            {
                if (body.Contains(c.ToString()))
                {
                    count2++;
                }
            }

            if (count1 > count2)
            {
                return title1;
            }
            return title2;
        }

        /// <summary>
        /// 格式化标签，剔除匹配标签中的回车符
        /// </summary>
        /// <param name="match"></param>
        /// <returns></returns>
        private static string FormatTag(Match match)
        {
            StringBuilder sb = new StringBuilder();
            foreach (var ch in match.Value)
            {
                if (ch == '\r' || ch == '\n')
                {
                    continue;
                }
                sb.Append(ch);
            }
            return sb.ToString();
        }

        /// <summary>
        /// 获取时间
        /// </summary>
        /// <param name="html"></param>
        /// <returns></returns>
        private static string GetTitle(string html)
        {
            //默认标题
            string titleFilter = @"<title>[\s\S]*?</title>";
            string clearFilter = @"<.*?>";

            string title = "";
            Match match = Regex.Match(html, titleFilter, RegexOptions.IgnoreCase);
            if (match.Success)
            {
                title = Regex.Replace(match.Groups[0].Value, clearFilter, "");
            }

            //尝试1： 正文的标题一般在h1中，比title中的标题更干净
            string h1Filter = @"<h1.*?>.*?</h1>";
            match = Regex.Match(html, h1Filter, RegexOptions.IgnoreCase);
            if (match.Success)
            {
                string h1 = Regex.Replace(match.Groups[0].Value, clearFilter, "");
                if (!String.IsNullOrEmpty(h1) && title.StartsWith(h1))
                {
                    return h1;
                }
            }

            title = title.Split(new char[] { '-', '_' }, StringSplitOptions.RemoveEmptyEntries)[0];
            title = title.Replace("\r\n", "").Trim();
            return title;
        }

        /// <summary>
        /// 获取文章发布日期
        /// </summary>
        /// <param name="html"></param>
        /// <returns></returns>
        private static DateTime GetPublishDate(string html)
        {
            // 过滤html标签，防止标签对日期提取产生影响
            string text = Regex.Replace(html, "(?is)<.*?>", "");
            Match match = Regex.Match(
                text,
                @"((\d{4}|\d{2})(\-|\/)\d{1,2}\3\d{1,2})(\s?\d{2}:\d{2})?|(\d{4}年\d{1,2}月\d{1,2}日)(\s?\d{2}:\d{2})?",
                RegexOptions.IgnoreCase);

            DateTime result = new DateTime(1900, 1, 1);
            if (match.Success)
            {
                try
                {
                    string dateStr = "";
                    for (int i = 0; i < match.Groups.Count; i++)
                    {
                        dateStr = match.Groups[i].Value;
                        if (!String.IsNullOrEmpty(dateStr))
                        {
                            break;
                        }
                    }
                    // 对中文日期的处理
                    if (dateStr.Contains("年"))
                    {
                        StringBuilder sb = new StringBuilder();
                        foreach (var ch in dateStr)
                        {
                            if (ch == '年' || ch == '月')
                            {
                                sb.Append("/");
                                continue;
                            }
                            if (ch == '日')
                            {
                                sb.Append(' ');
                                continue;
                            }
                            sb.Append(ch);
                        }
                        dateStr = sb.ToString();
                    }
                    result = Convert.ToDateTime(dateStr);
                }
                catch (Exception)
                {
                }
                if (result.Year < 1900)
                {
                    result = new DateTime(1900, 1, 1);
                }
            }
            return result;
        }

        /// <summary>
        /// 从body标签文本中分析正文内容
        /// </summary>
        /// <param name="bodyText">只过滤了script和style标签的body文本内容</param>
        /// <param name="content">返回文本正文，不包含标签</param>
        /// <param name="contentWithTags">返回文本正文包含标签</param>
        private static void GetContent(string bodyText, out string content, out string title, out string contentWithTags)
        {
            string[] orgLines = null; // 保存原始内容，按行存储
            string[] lines = null; // 保存干净的文本内容，不包含标签
            title = "";
            orgLines = bodyText.Split('\n');
            lines = new string[orgLines.Length];
            // 去除每行的空白字符,剔除标签
            for (int i = 0; i < orgLines.Length; i++)
            {
                string lineInfo = orgLines[i];
                // 处理回车，使用[crlf]做为回车标记符，最后统一处理
                lineInfo = Regex.Replace(lineInfo, "(?is)</p>|<br.*?/>", "[crlf]");
                lines[i] = Regex.Replace(lineInfo, "(?is)<.*?>", "").Trim();
            }

            StringBuilder sb = new StringBuilder();
            StringBuilder orgSb = new StringBuilder();

            int preTextLen = 0; // 记录上一次统计的字符数量
            int startPos = -1; // 记录文章正文的起始位置
            for (int i = 0; i < lines.Length - _depth; i++)
            {
                int len = 0;
                for (int j = 0; j < _depth; j++)
                {
                    len += lines[i + j].Length;
                }

                if (startPos == -1) // 还没有找到文章起始位置，需要判断起始位置
                {
                    if (preTextLen > _limitCount && len > 0) // 如果上次查找的文本数量超过了限定字数，且当前行数字符数不为0，则认为是开始位置
                    {
                        // 查找文章起始位置, 如果向上查找，发现2行连续的空行则认为是头部
                        int emptyCount = 0;
                        for (int j = i - 1; j > 0; j--)
                        {
                            if (String.IsNullOrEmpty(lines[j]))
                            {
                                emptyCount++;
                            }
                            else
                            {
                                emptyCount = 0;
                            }
                            if (emptyCount == _headEmptyLines)
                            {
                                startPos = j + _headEmptyLines;
                                break;
                            }
                        }
                        // 如果没有定位到文章头，则以当前查找位置作为文章头
                        if (startPos == -1)
                        {
                            startPos = i;
                        }
                        // 填充发现的文章起始部分
                        if (startPos > 2)
                        {
                            title = lines[startPos - 3];
                        }
                        for (int j = startPos; j <= i; j++)
                        {
                            sb.Append(lines[j]);
                            orgSb.Append(orgLines[j]);
                        }
                    }
                }
                else
                {
                    //if (len == 0 && preTextLen == 0)    // 当前长度为0，且上一个长度也为0，则认为已经结束
                    if (len <= _endLimitCharCount && preTextLen < _endLimitCharCount) // 当前长度为0，且上一个长度也为0，则认为已经结束
                    {
                        if (!_appendMode)
                        {
                            break;
                        }
                        startPos = -1;
                    }
                    sb.Append(lines[i]);
                    orgSb.Append(orgLines[i]);
                }
                preTextLen = len;
            }

            string result = sb.ToString();
            // 处理回车符，更好的将文本格式化输出
            content = result.Replace("[crlf]", Environment.NewLine);
            content = content.Replace("&nbsp;", "");
            content = content.Replace("&ldquo;", "");
            content = content.Replace("&rdquo;", "");
            content = System.Web.HttpUtility.HtmlDecode(content);
            // 输出带标签文本
            contentWithTags = orgSb.ToString();
        }
    }
}

File: Iveely.Framework/src/com/iveely/framework/net/Html2Article.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.iveely.framework.net;

import com.iveely.framework.java.RefObject;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.apache.log4j.Logger;

/**
 * Turn the page source code documentation.
 *
 * @author liufanping@iveely.com
 * @date 2014-10-25 22:24:51
 */
public class Html2Article {

    /**
     * Regex filters: regular expressions, was asked to replace text.
     */
    private static final String[][] filters
            = {
                new String[]{"(?is)<script.*?>.*?</script>", ""},
                new String[]{"(?is)<style.*?>.*?</style>", ""},
                new String[]{"(?is)<!--.*?-->", ""},
                new String[]{"(?is)</a>", "</a>\n"}
            };

    /**
     * Append mode.
     */
    private static boolean appendMode = false;

    /**
     * After whether to use append mode, the default is false to use append
     * mode, all text will meet the filter criteria extracted.
     *
     * @return
     */
    public static boolean getAppendMode() {
        return appendMode;
    }

    /**
     * Set append mode.
     *
     * @param value
     */
    public static void setAppendMode(boolean value) {
        appendMode = value;
    }

    /**
     * In depth analysis of the line, the default is 6.
     */
    private static int depth = 6;

    /**
     * Get depth.
     *
     * @return
     */
    public static int getDepth() {
        return depth;
    }

    /**
     * Set depth.
     *
     * @param value
     */
    public static void setDepth(int value) {
        depth = value;
    }

    /**
     * Limit the number of characters, when the number reaches a limited number
     * of text analysis is considered 180 characters into the text content of
     * the default number.
     */
    private static int limitCount = 180;

    /**
     * Get limit count.
     *
     * @return
     */
    public static int getLimitCount() {
        return limitCount;
    }

    /**
     * Set limit count.
     *
     * @param value
     */
    public static void setLimitCount(int value) {
        limitCount = value;
    }

    /**
     * Determining the body of the article head, look upward, consecutive empty
     * lines reach headEmptyLines, then stop searching.
     */
    private static final int headEmptyLines = 2;

    /**
     * Used to determine the number of characters in the article ends.
     */
    private static final int endLimitCharCount = 20;

    /**
     * Logger(log4j)
     */
    private final static Logger logger = Logger.getLogger(Html2Article.class.getName());

    /**
     * Get information from a given text Html original text.
     *
     * @param code
     * @return
     */
    public static String getContent(String code) {
        if (code == null) {
            return "";
        }
        if (stringNumbers(0, code, "\n") < 10) {
            code = code.replace(">", ">\n").toLowerCase();
        }
        String body;
        String bodyFilter = "<body.*?</body>";
        Pattern pattern = Pattern.compile(bodyFilter, Pattern.DOTALL);
        Matcher matcher = pattern.matcher(code);
        StringBuilder sb = new StringBuilder();
        if (matcher.find()) {
            sb.append(matcher.group());
        }
        body = sb.toString();
        for (String[] filter : filters) {
            body = body.replaceAll(filter[0], filter[1]);
        }

        String content = null;
        String contentWithTags = null;
        RefObject<String> tempRef_content = new RefObject<>(content);
        RefObject<String> tempRef_contentWithTags = new RefObject<>(contentWithTags);
        getContent(body, tempRef_content, tempRef_contentWithTags);
        content = tempRef_content.argvalue.trim();
        if (content.length() < 20) {
            content = code.toLowerCase().replaceAll("<script(?:[^<]++|<(?!/script>))*+</script>", "").replaceAll("<style(?:[^<]++|<(?!/style>))*+</style>", "").replaceAll("<.*?>", "").trim();
        } else {
            content = content.toLowerCase().replaceAll("<script(?:[^<]++|<(?!/script>))*+</script>", "").replaceAll("<style(?:[^<]++|<(?!/style>))*+</style>", "").replaceAll("<.*?>", "").trim();
        }
        return content;
    }

    /**
     * Get meta description.
     *
     * @param code
     * @return
     */
    public static String getMetaDesc(String code) {
        String patternVal = "<meta.*?>";
        Pattern pattern = Pattern.compile(patternVal, Pattern.DOTALL);
        Matcher matcher = pattern.matcher(code);
        while (matcher.find()) {
            String temp = matcher.group();
            if (temp.contains("description")) {
                temp = temp.replace("content=", "").replace("<meta", "").replace("\"", "").replace("/>", "").trim().replace("name=description", "").replace(">", "");
                return temp;
            }
        }
        return "";
    }

    /**
     * Formatting tags, excluding matching label carriage.
     *
     * @param match
     * @return
     */
    private static String formatTag(Matcher match) {
        StringBuilder sb = new StringBuilder();
        for (char ch : match.group(0).toCharArray()) {
            if (ch == '\r' || ch == '\n') {
                continue;
            }
            sb.append(ch);
        }
        return sb.toString();
    }

    /**
     * Get title.
     *
     * @param html
     * @param backupTitle
     * @return
     */
    public static String getTitle(String html, String backupTitle) {
        String titleFilter = "<title>[\\s\\S]*?</title>";
        String h1Filter = "<h1.*?>.*?</h1>";
        String clearFilter = "<.*?>";
        String title = "";
        Pattern pattern = Pattern.compile(titleFilter, Pattern.DOTALL);
        Matcher matcher = pattern.matcher(html.toLowerCase());
        if (matcher.find()) {
            title = matcher.group().replaceAll(clearFilter, "").trim();
        }
        pattern = Pattern.compile(h1Filter, Pattern.DOTALL);
        matcher = pattern.matcher(html);
        if (matcher.find()) {
            String h1 = matcher.group(0).replace(clearFilter, "");
            if (!(h1 == null || h1.equals("")) && title.startsWith(h1)) {
                title = h1.split("-| ")[0].trim();
            }
        }
        return title;
    }

    /**
     * get publish date.
     *
     * @param html
     * @return
     */
    public static String getPublishDate(String html) {
        Pattern pattern = Pattern.compile("(?is)<.*?>", Pattern.DOTALL);
        Matcher matcher = pattern.matcher(html);
        String text = matcher.replaceAll("");
        pattern = Pattern.compile("((\\d{4}|\\d{2})(\\-|\\/)\\d{1,2}\\3\\d{1,2})(\\s?\\d{2}:\\d{2})?|(\\d{4}年\\d{1,2}月\\d{1,2}日)(\\s?\\d{2}:\\d{2})?", Pattern.DOTALL);
        matcher = pattern.matcher(text);
        if (matcher.find()) {
            try {
                String dateStr = "";
                for (int i = 0; i < matcher.groupCount(); i++) {
                    dateStr = matcher.group(i);
                    if (!(dateStr == null || dateStr.equals(""))) {
                        break;
                    }
                }
                if (dateStr.contains("年")) {
                    StringBuilder sb = new StringBuilder();
                    for (char ch : dateStr.toCharArray()) {
                        if (ch == '年' || ch == '月') {
                            sb.append("/");
                            continue;
                        }
                        if (ch == '日') {
                            sb.append(' ');
                            continue;
                        }
                        sb.append(ch);
                    }
                    dateStr = sb.toString();
                }
                return dateStr;
            } catch (RuntimeException ex) {
                System.out.println(ex);
            }
        }
        return "";
    }

    /**
     * Analysis of the text content from the body text of the label.
     *
     * @param bodyText
     * @param content
     * @param contentWithTags
     */
    private static void getContent(String bodyText, RefObject<String> content, RefObject<String> contentWithTags) {
        String[] orgLines;
        String[] lines;
        orgLines = bodyText.split("[\\n]", -1);
        lines = new String[orgLines.length];
        for (int i = 0; i < orgLines.length; i++) {
            String lineInfo = orgLines[i];
            Pattern pattern = Pattern.compile("(?is)</p>|<br.*?/>|</H1>|</H2>|</H3>");
            Matcher matcher = pattern.matcher(lineInfo);
            lineInfo = matcher.replaceAll("[crlf]");
            pattern = Pattern.compile("(?is)<.*?>");
            matcher = pattern.matcher(lineInfo);
            lines[i] = matcher.replaceAll("");
        }
        if (lines.length > 1000) {
            StringBuilder contentBuffer = new StringBuilder();
            for (String line : lines) {
                String temp = line.trim();
                if (temp.length() > 0) {
                    contentBuffer.append(temp).append(" ");
                }
            }
            content.argvalue = contentBuffer.toString().replace("[crlf]", "\n").replace(" 　　", "\n").replace("&nbsp;", "\n");
            return;
        }
        StringBuilder sb = new StringBuilder();
        StringBuilder orgSb = new StringBuilder();
        int preTextLen = 0;
        int startPos = -1;
        for (int i = 0; i < lines.length - depth; i++) {
            int len = 0;
            for (int j = 0; j < depth; j++) {
                len += lines[i + j].length();
            }
            if (startPos == -1) {
                if (preTextLen > limitCount && len > 0) {
                    int emptyCount = 0;
                    for (int j = i - 1; j > 0; j--) {
                        if (lines[j] == null || lines[j].equals("")) {
                            emptyCount++;
                        } else {
                            emptyCount = 0;
                        }
                        if (emptyCount == headEmptyLines) {
                            startPos = j + headEmptyLines;
                            break;
                        }
                    }
                    if (startPos == -1) {
                        startPos = i;
                    }
                    for (int j = startPos; j <= i; j++) {
                        sb.append(lines[j]);
                        orgSb.append(orgLines[j]);
                    }
                }
            } else {
                if (len <= endLimitCharCount && preTextLen < endLimitCharCount) {
                    if (!appendMode) {
                        break;
                    }
                    startPos = -1;
                }
                sb.append(lines[i]);
                orgSb.append(orgLines[i]);
            }
            preTextLen = len;
        }
        String result = sb.toString();
        content.argvalue = result.replace("[crlf]", "\n").replace(" 　　", "\n").replace("&nbsp;", "\n");
        contentWithTags.argvalue = orgSb.toString();
    }

    private static int stringNumbers(int defaultCount, String source, String childString) {
        if (!source.contains(childString)) {
            return defaultCount + 0;
        } else if (source.contains(childString)) {
            defaultCount++;
            return stringNumbers(defaultCount, source.substring(source.indexOf(childString) + childString.length()), childString);
        }
        return defaultCount;
    }

    /**
     * Get logo's url.
     *
     * @param fromUrl
     * @param htmlCode
     * @return
     */
    public static String getLogoUrl(String fromUrl, String htmlCode) {
        final String regex = "<link.*?>";
        String url = "";
        if (htmlCode == null) {
            return url;
        }
        final Pattern pt = Pattern.compile(regex, Pattern.DOTALL);
        final Matcher mt = pt.matcher(htmlCode);
        while (mt.find()) {
            final Matcher urlMatcher = Pattern.compile("href=.*?>", Pattern.DOTALL).matcher(mt.group());
            String childUrl;
            while (urlMatcher.find()) {
                childUrl = urlMatcher.group();
                if (childUrl.toLowerCase().contains(".ico")) {
                    String[] urlComs = childUrl.split(" ");
                    String urlCom = urlComs[0].replaceAll("href=|>", "");
                    if (urlCom.length() > 5) {
                        url = joinUrl(fromUrl, urlCom);
                        if (url.endsWith("/")) {
                            url = url.substring(0, url.length() - 1);
                        }
                        break;
                    }
                }
            }
        }
        if (url.length() < 5) {
            url = "";
        }
        return url;
    }

    /*
     Join url.
     */
    private static String joinUrl(String curl, String file) {
        file = file.replace("\"", "");
        if (file.toLowerCase().startsWith("http")) {
            return file;
        }
        URL url;
        String q = "";
        try {
            url = new URL(new URL(curl), file);
            q = url.toExternalForm();
        } catch (MalformedURLException e) {
            logger.error(e);
        }
        if (q.contains("#")) {
            q = q.replaceAll("^(.+?)#.*?$", "$1");
        }
        return q.replace("/../", "/");
    }
}
